concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -116,30 +125,122 @@ sectionHeaderLength :: Word64
 sectionHeaderLength = fromIntegral $ BS.length $ encode placeholderSectionHeader
 
 -- |Open a file handle for writing a chunk. If the file with the specified name already exists, a
--- new name is chosen by putting or incrementing a version number in its extension. It is expected
--- that an unversioned filename only has a single extension.
-initialHandle :: FilePath -> IO (FilePath, Handle)
-initialHandle p = do
+-- boolean flag indicates whether the file should be overwritten or whether a new name should be
+-- chosen by adding to or incrementing a version number in its extension. It is expected that an
+-- unversioned filename only has a single extension.
+initialHandle :: Bool -> FilePath -> IO (FilePath, Handle)
+initialHandle overwriteExistingChunk p = do
     createDirectoryIfMissing True (takeDirectory p)
     chunkExists <- doesFileExist p
-    if chunkExists
-        then initialHandle (replaceExtensions p newExt)
+    if chunkExists && not overwriteExistingChunk
+        then do
+            newExt <- getNewExt
+            initialHandle overwriteExistingChunk (replaceExtensions p newExt)
         else do
             hdl <- openBinaryFile p WriteMode
             return (p, hdl)
   where
+    -- The old, potentially unversionsioned, file extension
+    oldExt :: String
     oldExt = takeExtensions p
-    newExt =
+    -- The next versioned file extension
+    getNewExt :: IO String
+    getNewExt =
         if oldExt == takeExtension p
-            then ".2" ++ oldExt
-            else
-                (\x -> "." ++ x ++ ".dat")
-                    . show
-                    . (+ (1 :: Int))
-                    . read
-                    . (!! 1)
-                    . splitOn "."
-                    $ oldExt
+            then return $ ".2" ++ oldExt
+            else case tryGetVersionNumber (T.pack oldExt) of
+                Left err -> fail err
+                Right v -> return $ "." <> show (v + 1) <> ".dat"
+    -- Given a file name attempt to get its version number
+    tryGetVersionNumber :: T.Text -> Either String Integer
+    tryGetVersionNumber = AP.parseOnly $ do
+        AP.skipMany $ AP.skip (AP.inClass "a-zA-Z0-9-")
+        _ <- AP.char '.'
+        AP.decimal
+
+-- |Data type used to represent a line with genesis block hash in the block index file.
+type BlockIndexGenesisHash = BlockHash
+
+-- |Data type used to represent a line with chunk information in the block index file.
+data BlockIndexChunkInfo = BlockIndexChunkInfo
+    { filename :: T.Text,
+      genesisIndex :: GenesisIndex,
+      blockHeightFirst :: BlockHeight,
+      blockHeightLast :: BlockHeight
+    }
+    deriving (Show)
+
+-- |Data type used to represent the block index file.
```

Comment:
This documenation is not very useful. It would be better to state the intent. 

"An index file is a list of lists of chunks. Each list of chunks is associated with a hash of the genesis to which all the blocks belong to."

Or some such.